engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/Operator.scala

```diff
@@ -144,6 +144,19 @@ object Operator {
     /** @inheritdoc */
     override def children: List[IR] = List(left, operator, right)
 
+    /** @inheritdoc */
+    override def withNewChildren(newChildren: List[IR]): IR = {
+      newChildren match {
```

Comment:
I thought this should be implemented in terms of `mapExpressions` - e.g.
```scala
mapExpressions { ir =>
   val index = children().indexOf(ir)
   newChildren(index)
}
```
or at least it might be a good default implementation for `ir.Expression`.

Reply:
That would be nice. Only if `children().indexOf(ir)` would not always return -1. Let's make this simple for now and just implement `withNewChildren` whenever it is required.

Reply:
> implement `withNewChildren` whenever it is required.

E.g. in every `IR` sub type? Hm. Another reason to **generate IR by annotation processor**...

> `children().indexOf(ir)` always return(s) `-1`

Why does it do so? For example `Expression.Block` has [mapExpression](https://github.com/enso-org/enso/blob/develop/engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/Expression.scala#L144) that operates on `expressions` and `returnValue`. Its [children() implementation](https://github.com/enso-org/enso/blob/develop/engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/Expression.scala#L168) returns `expressions` and `returnValue` - e.g. each `ir` passed into the callback function must be in the `children()` list!

Reply:
1) `mapExpressions` is implemented separately for each IR type and it works **recursively**.
2) `children` is implemented separately for each IR type as well.

So yes, generating these IR-manipulation methods by an annotation processor would be very valuable. Until we have this functionality, let's just implement `withNewChildren` as a complementary method for `children` for each IR type that needs it. And test it (in 40252a22df7d6719cf572890aa47d0bb22609a1f)

Reply:
My [experiments don't show any traces of recursion](https://github.com/enso-org/enso/issues/10981#issuecomment-2390634772) in `mapExpressions` as far as I can tell.

Reply:
Conclusion: the problem isn't recursion. Continuation of the discussion about `mapExpressions` is [here](https://github.com/enso-org/enso/pull/11191#discussion_r1787612365)